tests/benchmark/utils/s2n_negotiate_client_benchmark.cc

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tests/benchmark/utils/s2n_negotiate_client_benchmark.h"
+#include "tests/benchmark/utils/shared_info.h"
+#include "utils/s2n_safety.h"
+#include "utils/s2n_random.h"
+#include <openssl/err.h>
+#include <openssl/crypto.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <iostream>
+#include <poll.h>
+#include <vector>
+#include <stdlib.h>
+
+extern "C" {
+#include "bin/common.h"
+#include "error/s2n_errno.h"
+#include "tls/s2n_connection.h"
+}
+
+static int setup_socket(struct addrinfo hints, struct addrinfo *ai_list, struct addrinfo *ai) {
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+    int add = 0;
+
+    if ((add = getaddrinfo(host, port, &hints, &ai_list)) != 0) {
+        fprintf(stderr, "getaddrinfo error: %s\nError: %d\nErrno: %s\n", gai_strerror(add), add, strerror(errno));
+        return -1;
+    }
+
+    bool connected = false;
+    while (!connected) {
+        for (ai = ai_list; ai != NULL; ai = ai->ai_next) {
+            if ((sockfd = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol)) == -1) {
+                continue;
+            }
+            if (connect(sockfd, ai->ai_addr, ai->ai_addrlen) == -1) {
+                close(sockfd);
+                continue;
+            }
+            connected = true;
+            if (DEBUG_PRINT) {
+                printf("Connected to s2nd\n");
+            }
+            break;
+        }
+    }
+
+    freeaddrinfo(ai_list);
+    GUARD_EXIT(sockfd, "Socket setup failed\n");
+    return 0;
+}
+
+static void client_handshake(benchmark::State& state, bool warmup, struct s2n_connection *conn) {
+    GUARD_EXIT(s2n_set_server_name(conn, host), "Error setting server name");
+    GUARD_EXIT(s2n_connection_set_fd(conn, sockfd), "Error setting file descriptor");
+
+    if (benchmark_negotiate(conn, sockfd, state, warmup) != S2N_SUCCESS) {
+        state.SkipWithError("Negotiate Failed\n");
+        if (DEBUG_PRINT) {
+            printf("Error in negotiate!\n");
+        }
+    }
+
+    if (DEBUG_PRINT) {
+        printf("Connected to %s:%s\n", host, port);
+    }
+
+    GUARD_EXIT(s2n_connection_free_handshake(conn), "Error freeing handshake memory after negotiation");
```

Comment:
I don't think you need to do this `s2n_connection_wipe` already wipes and reallocates everything in the handshake.